workers/cpp_physics/physics.cc

```diff
@@ -0,0 +1,104 @@
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <string>
+#include <thread>
+#include <cmath>
+
+#include <improbable/worker.h>
+#include <improbable/view.h>
+
+#include <improbable/standard_library.h>
+#include <client_data.h>
+
+using ComponentRegistry = worker::Components<improbable::Position, improbable::Metadata>;
+
+const int kErrorExitStatus = 1;
+const std::uint32_t kGetOpListTimeoutInMilliseconds = 100;
+
+int main(int argc, char** argv) {
+    auto print_usage = [&]() {
+        std::cout << "Usage: Physics <hostname> <port> <worker_id>" << std::endl;
+        std::cout << std::endl;
+        std::cout << "Connects to SpatialOS" << std::endl;
+        std::cout << "    <hostname>      - hostname of the receptionist to connect to." << std::endl;
+        std::cout << "    <port>          - port to use if connecting through the receptionist." << std::endl;
+        std::cout << "    <worker_id>     - (optional) name of the worker assigned by SpatialOS." << std::endl;
+        std::cout << std::endl;
+    };
+
+    if (argc != 4) {
+        print_usage();
+        return kErrorExitStatus;
+    }
+
+    worker::EntityId entityId = 1;
+
+    worker::ConnectionParameters parameters;
+    parameters.WorkerType = "physics";
+    parameters.Network.ConnectionType = worker::NetworkConnectionType::kTcp;
+    parameters.Network.UseExternalIp = false;
+    worker::Connection connection = worker::Connection::ConnectAsync(
+        ComponentRegistry{}, std::string{argv[1]}, atoi(argv[2]), argv[3], parameters).Get();
+
+    if (connection.GetConnectionStatusCode() != worker::ConnectionStatusCode::kSuccess) {
+        std::cerr << "Failed to connection the receptionist." << std::endl;
+        std::cerr << "Reason: " << connection.GetConnectionStatusDetailString() << std::endl;
+        return kErrorExitStatus;
+    }
+    std::cout << "Successfully connected using the Receptionist" << std::endl;
+
+    worker::View view{ComponentRegistry{}};
+    bool is_connected = true;
+
+    view.OnDisconnect([&](const worker::DisconnectOp& op) {
+        std::cerr << "[disconnect] " << op.Reason << std::endl;
+        is_connected = false;
+    });
+    view.OnLogMessage([&](const worker::LogMessageOp& op) {
+        if (op.Level == worker::LogLevel::kFatal) {
+            std::cerr << "Fatal error: " << op.Message << std::endl;
+            is_connected = false;
+        } else {
+            std::cout << "[remote] " << op.Message << std::endl;
+        }
+    });
+    view.OnCommandRequest<sample::Login::Commands::TakeControl>([&](const worker::CommandRequestOp<sample::Login::Commands::TakeControl>& op) {
+        connection.SendLogMessage(worker::LogLevel::kInfo, "Physics", "Assigning ClientData component to " + op.CallerAttributeSet[1]);
```

Comment:
`op.CallerAttributeSet[1]`, is it still guaranteed to have two attributes?

Reply:
Yes.